### PR TITLE
In-place rebinning in the instrument view

### PR DESCRIPTION
**Pull request by MathieuTillet**
**Description of work.**
This PR adds a panel inside the pick tab of the instrument view to allow the user to rebin the workspace without leaving the instrument viewer. Since it replaces the workspace, a couple of bugs have been solved at the same time.

Now, the pick tab also shows the integration over the entire detector if in shape drawing mode and no shape is currently drawn, instead of just showing nothing.

This PR relies on changes from https://github.com/mantidproject/mantid/pull/31754 , do not merge it before this one has been merged first.

**To test:**
- Load a file - preferably an event file like `ExternalData/Testing/Data/SystemTest/CNCS_7860_event.nxs` . 
- Open the instrument viewer, and in the pick tab, change the mode to shape drawing, shape editing or zoom, and check the integration appears and looks correct.
- In the rebin panel under the integration plot, try a few rebin values, and check the plot is correctly updated.
- Finally, checks that saving as histogram does produce a workspace without events.

WIP for #31747 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by gvardany** [comment]
@MathieuTillet the dependent PR got merged, can you resolve this?

File: qt/widgets/instrumentview/src/InstrumentWidgetPickTab.cpp [hunk 1]
```diff
@@ -100,6 +97,33 @@ InstrumentWidgetPickTab::InstrumentWidgetPickTab(InstrumentWidget *instrWidget)
   m_plot = new MiniPlot(this);
   connect(m_plot, SIGNAL(showContextMenu()), this, SLOT(plotContextMenu()));
 
+  // set up the rebin tab
+  m_rebin = new QWidget(this);
+  QGridLayout *rebinLayout = new QGridLayout(m_rebin);
+  m_rebinParams = new QLineEdit(m_rebin);
+  m_rebinParams->setToolTip("The parameters to use for the new binning, a comma separated list of first bin boundary, "
+                            "width, last bin boundary. See Rebin doc for details.");
+  m_rebinUseReverseLog = new QCheckBox("UseReverseLogarithmic", m_rebin);
```
Comment (gvardany) [inline qt/widgets/instrumentview/src/InstrumentWidgetPickTab.cpp:106]: Please change the name of the checkbox to "Use reverse logarithmic" to be consistent in style with the other. You should keep, however, the name of the property as is.

**Review by gvardany — COMMENTED** [review]
Code-wise looks alright, I'd just suggest to add unit tests to the vector helper as well. 
I'd add assertion in the tests that all the bin widths are in the expected order; that is, monotonically increasing for logarithmic, monotonically decreasing for reverse logarithmic and power. Will do functional tests shortly.

File: Framework/Algorithms/src/Rebin.cpp [hunk 2]
```diff
@@ -89,31 +139,30 @@ void Rebin::init() {
...
+                  "Ignore errors related to zero/negative bin widths in input/output workspaces. When ignored, the "
+                  "signal and errors are set to zero");
+
+  declareProperty("UseReverseLogarithmic", false,
+                  "For logarithmic intervals, the splitting starts from the end and go back to the start, ie the bins "
+                  "are bigger at the start and exponentially reduces until they reach the end. For these bins, the "
+                  "FullBinsOnly flag is ignored.");
+
+  declareProperty("Power", EMPTY_DBL(),
+                  "Splits the interval in bins whose width is width / (i ^ power). Power must be between 0 and 1.");
```
Comment (gvardany) [inline Framework/Algorithms/src/Rebin.cpp:165]:
```suggestion
                  "Splits the interval in bins with actual width equal to the requested width / (i ^ power). Power must be between 0 and 1.");
```

File: Framework/Algorithms/src/Rebin.cpp [hunk 3]
```diff
@@ -89,31 +139,30 @@ void Rebin::init() {
...
   declareProperty("IgnoreBinErrors", false,
-                  "Ignore errors related to "
-                  "zero/negative bin widths in "
-                  "input/output workspaces. When ignored, the signal and "
-                  "errors are set to zero");
+                  "Ignore errors related to zero/negative bin widths in input/output workspaces. When ignored, the "
+                  "signal and errors are set to zero");
+
+  declareProperty("UseReverseLogarithmic", false,
+                  "For logarithmic intervals, the splitting starts from the end and go back to the start, ie the bins "
```
Comment (gvardany) [inline Framework/Algorithms/src/Rebin.cpp:160]: There are spelling mistakes here:
```suggestion
                  "For logarithmic intervals, the splitting starts from the end and goes back to the start, ie the bins "
```

File: Framework/Algorithms/src/Rebin.cpp [hunk 4]
```diff
@@ -89,31 +139,30 @@ void Rebin::init() {
...
+                  "are bigger at the start and exponentially reduces until they reach the end. For these bins, the "
```
Comment (gvardany) [inline Framework/Algorithms/src/Rebin.cpp:161]:
```suggestion
                  "are bigger at the start getting exponentially reduced until they reach the end. For these bins, the "
```

**Comment by gvardany** [comment]
Functionality still works to spec.

File: qt/widgets/instrumentview/src/InstrumentWidgetPickTab.cpp (hunk 1 above)
Comment (MathieuTillet) [reply qt/widgets/instrumentview/src/InstrumentWidgetPickTab.cpp:106]: I was being consistent with Rebin parameters, but noted.

Comment (gvardany) [reply qt/widgets/instrumentview/src/InstrumentWidgetPickTab.cpp:106]: Well but the other checkbox you don't put at all the actual property name. So either we take both property names as is from the Rebin, so (PreserveEvents and UseReverseLogarithmic), or if we rewrite our own labels, we should do for both to make them consistent.

File: Framework/Algorithms/test/RebinTest.h [hunk 5]
```diff
@@ -574,6 +881,25 @@ class RebinTest : public CxxTest::TestSuite {
...
+
+  bool checkBinWidthMonotonic(MatrixWorkspace_sptr ws, bool reverse = false, bool ignoreLastBin = false) {
+    size_t binEdgesTotal = ws->blocksize();
+    if (ignoreLastBin)
+      binEdgesTotal--;
+    auto binEdges = ws->binEdges(0);
+    double lastBinSize = binEdges[1] - binEdges[0];
+    for (size_t i = 1; i < binEdgesTotal; ++i) {
+      double currentBinSize = binEdges[i + 1] - binEdges[i];
+      std::cout << currentBinSize << std::endl;
```
Comment (gvardany) [inline Framework/Algorithms/test/RebinTest.h:893]: Good tests added, one last leftover, otherwise looks good.

**Review by gvardany — DISMISSED** [review]
Good to go.

**Review by gvardany — APPROVED** [review]

**Pull request merged**